# Advanced Programming - Assignment 1

## 1. Extensible Graph Generator – OOP Design


In [65]:
import random

class Nodes():
    """The class takes a number of nodes as argument.
    It then creates a dictionary containing the number of nodes as keys and empty lists as values.
    The empty lists will later be used to add the edges
    """
    def create_nodes(self, n_nodes):
        nodes = {}
        for i in range(1, n_nodes+1):
            nodes[i] = []
        return nodes

class DirectedEdges():
    """The edges are determined for every node, based on a fixed probability.
    The edges are represented as the values in the dictionary.
    The edges are directed, so an edge from node 1 to node 4 does not mean node 4 has an edge with node 1.
    """
    def edges(self, n_nodes, node_dict):
        for i in range(1,n_nodes+1):
            for j in range(1, n_nodes+1):
                edge_prob = random.uniform(0, 1)
                if edge_prob > 0.5:
                    continue
                else:
                    if j == i:
                        continue
                    else:
                        node_dict[i].append(j)
        return node_dict
            
class UndirectedEdges():
    """The edges are determined for every node, based on a fixed probability.
    The edges are represented as the values in the dictionary.
    The edges are undirected, so an edge from node 1 to node 4 also means node 4 has an edge with node 1.
    """
    def edges(self, n_nodes, node_dict):
        for i in range(1,n_nodes+1):
            for j in range(1, n_nodes+1):
                edge_prob = random.uniform(0, 1)
                if edge_prob > 0.5:
                    continue
                else:
                    if j == i:
                        continue
                    else:
                        if i in node_dict[i] or j in node_dict[i]:
                            continue
                        else:
                            node_dict[i].append(j)
                            node_dict[j].append(i)
        return node_dict

class Graph():
    """The function requires to specify if the edges are directed or undirected and requires an amount of nodes.
    The class uses the nodes and edges classes to generate the graph and then returns it.
    """
    def __init__(self, edge_direction, n_nodes):
        self.edge = edge_direction
        self.n_nodes = n_nodes
        self.generate_nodes = {}
        self.edge_dict = {}
    def generate(self):
        node_initialization = Nodes()
        generate_nodes = node_initialization.create_nodes(self.n_nodes)
        self.edge_dict = self.edge.edges(self.n_nodes, generate_nodes)
        return self.edge_dict
    def __str__(self):
        return "Graph: {}".format(self.edge_dict.items())
    
x = Graph(DirectedEdges(), 10)
x.generate()
print(x)

print()
y = Graph(UndirectedEdges(), 10)
y.generate()
print(y)

Graph: dict_items([(1, [3, 5, 6, 7, 10]), (2, [1, 4, 5, 6, 8, 10]), (3, [2, 6, 9]), (4, [1, 2, 7, 10]), (5, [1, 4, 9]), (6, [2, 4, 8, 9, 10]), (7, [1, 2, 4, 8, 9, 10]), (8, [2, 4, 5, 7, 9, 10]), (9, [1, 4, 5, 6, 7, 8, 10]), (10, [2, 3, 6])])

Graph: dict_items([(1, [2, 6, 7, 9, 3, 4, 5, 8]), (2, [1, 3, 4, 6, 8, 10, 9]), (3, [2, 1, 4, 5, 6, 9, 7, 8]), (4, [2, 3, 1, 9, 7, 10]), (5, [3, 1, 7, 8, 9, 10, 6]), (6, [1, 2, 3, 5, 8, 9, 10]), (7, [1, 5, 3, 4, 10]), (8, [2, 5, 6, 1, 3]), (9, [1, 3, 4, 5, 6, 2, 10]), (10, [2, 5, 6, 9, 4, 7])])


## 2. Flexible Graph Generation Strategies

### A: 
Implement a Graph-Generator class hierarchy according to the above models that should
take a graph generation strategy as an argument, and return the generated graph.
Implement this in an OOP design.

In [66]:
import random

class Nodes():
    """The class takes a number of nodes as argument.
    It then creates a dictionary containing the number of nodes as keys and empty lists as values.
    The empty lists will later be used to add the edges
    """
    def create_nodes(self, n_nodes):
        nodes = {}
        for i in range(1, n_nodes+1):
            nodes[i] = []
        return nodes

class Strategy1():
    def edges(self, n_nodes, node_dict):
        return_dict = {}
        return return_dict

class Strategy2():
    def edges(self, n_nodes, node_dict):
       return_dict = {}
       return return_dict

class Graph():
    """The function requires to specify if the edges are directed or undirected and requires an amount of nodes.
    The class uses the nodes and strategy classes to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = {}
        self.strategy_dict = {}
    def generate(self):
        node_initialization = Nodes()
        generate_nodes = node_initialization.create_nodes(self.n_nodes)
        self.strategy_dict = self.strategy.edges(self.n_nodes, generate_nodes)
        return self.strategy_dict
    def __str__(self):
        return "Graph: {}".format(self.strategy_dict.items())
    
x = Graph(Strategy1(), 10)
x.generate()
print(x)

print()
y = Graph(Strategy2(), 10)
y.generate()
print(y)

Graph: dict_items([])

Graph: dict_items([])


Changes: the directed or undirected classes have been removed. 
Instead, a specific strategy, which by itself is either directed or undirected, is specified.
I interpreted the question as follows: I only changed the hierarchy of the classes in order to implement a graph generation strategy
The code does not yet implement any strategies, as this is asked in sub-question b and c.


### B:
Implement a Erdos-Renyi generation strategy for directed and undirected graphs (each).

In [67]:
import random

class Nodes():
    """The class takes a number of nodes as argument.
    It then creates a dictionary containing the number of nodes as keys and empty lists as values.
    The empty lists will later be used to add the edges
    """
    def create_nodes(self, n_nodes):
        nodes = {}
        for i in range(1, n_nodes+1):
            nodes[i] = []
        return nodes

class ErdosRenyiDirected():
    """The class takes a probability as argument.
    It uses the dictionary from the nodes class and adds the edges (values), if the probability threshold is exceeded,
    to the nodes (keys). 
    The graph is directed, so and edge between node 1 and 2 does not mean and edge between node 2 and 1.
    """
    def __init__(self, probability):
        self.probability = probability
    def edges(self, n_nodes, node_dict):
        for i in range(1,n_nodes+1):
            for j in range(1, n_nodes+1):
                edge_prob = random.uniform(0, 1)
                if edge_prob > self.probability:
                    continue
                else:
                    if j == i:
                        continue
                    else:
                        node_dict[i].append(j)
        return node_dict
    
class ErdosRenyiUndirected():
    """The class takes a probability as argument.
    It uses the dictionary from the nodes class and adds the edges (values), if the probability threshold is exceeded,
    to the nodes (keys). 
    The graph is undirected, so and edge between node 1 and 2 does automatically generates and edge between node 2 and 1.
    """
    def __init__(self, probability):
        self.probability = probability
    def edges(self, n_nodes, node_dict):
        for i in range(1,n_nodes+1):
            for j in range(1, n_nodes+1):
                edge_prob = random.uniform(0, 1)
                if edge_prob > self.probability:
                    continue
                else:
                    if j == i:
                        continue
                    else:
                        if i in node_dict[i] or j in node_dict[i]:
                            continue
                        else:
                            node_dict[i].append(j)
                            node_dict[j].append(i)
        return node_dict

class Graph():
    """The function requires to specify a strategy and requires an amount of nodes.
    The class uses the nodes and strategy classes to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = {}
        self.strategy_dict = {}
    def generate(self):
        node_initialization = Nodes()
        generate_nodes = node_initialization.create_nodes(self.n_nodes)
        self.strategy_dict = self.strategy.edges(self.n_nodes, generate_nodes)
        return self.strategy_dict
    def __str__(self):
        return "Graph: {}".format(self.strategy_dict.items())
    
x = Graph(ErdosRenyiDirected(0.75), 10)
x.generate()
print(x)
print()

y = Graph(ErdosRenyiUndirected(0.6), 10)
y.generate()
print(y)

Graph: dict_items([(1, [2, 3, 4, 7, 8, 9, 10]), (2, [1, 3, 5, 6, 8, 9, 10]), (3, [4, 5, 6, 8, 9, 10]), (4, [1, 2, 3, 6, 7, 8, 9, 10]), (5, [1, 2, 7, 8, 10]), (6, [2, 3, 7, 8, 9]), (7, [3, 4, 5, 6, 8, 9, 10]), (8, [2, 3, 4, 5, 7, 9, 10]), (9, [2, 3, 4, 6, 8, 10]), (10, [1, 2, 3, 5, 6, 7, 8, 9])])

Graph: dict_items([(1, [4, 5, 6, 7, 9, 10, 2, 3, 8]), (2, [1, 6, 8, 4, 9]), (3, [1, 5, 8, 9, 4, 6]), (4, [1, 2, 3, 5, 8, 9, 10, 6, 7]), (5, [1, 3, 4, 8, 9, 10, 6]), (6, [1, 2, 3, 4, 5, 8, 10, 7, 9]), (7, [1, 4, 6, 8, 10, 9]), (8, [2, 3, 4, 5, 6, 7, 1, 9, 10]), (9, [1, 3, 4, 5, 8, 2, 6, 7, 10]), (10, [1, 4, 5, 6, 7, 8, 9])])


### C:
Implement the Barabasi-Albert strategy for undirected graphs.

In [68]:
import random

class Nodes():
    """The class takes a number of nodes as argument.
    It then creates a dictionary containing the number of nodes as keys and empty lists as values.
    The empty lists will later be used to add the edges
    """
    def create_nodes(self, n_nodes):
        nodes = {}
        for i in range(1, n_nodes+1):
            nodes[i] = []
        return nodes

class BarabasiAlbertUndirected():
    """The function takes an amount of nodes to be added and an amount of edges (m) to be added for each new node.
    The graph is initialized with three nodes and some edges. For each new node and each new edge m: the node to which
    said edge is connected is determined by the amount of edges a node has with respect to the total amount of edges.
    If there is a 'dominant' node (most edges), then an edge is made between the new node and said dominant node. 
    If there are more then one nodes with the same amount of edges, chance decides with node gets an edge with the new node.
    """
    def __init__(self, m):
        self.m = m
    def edges(self, n_nodes, useless_dict):
        node_count = 4 #Starting number, since there are 3 inital nodes
        node_dict = {1:[2], 2:[1,3], 3:[2]} # the initial setup
        while node_count <= n_nodes+3:
                node_dict[node_count] = []
                for i in node_dict:
                    if i == node_count:
                        continue
                    else:
                        prob = random.uniform(0, 1)
                        total_edges = 0
                        for j in node_dict:
                            total_edges += len(node_dict[j])
                        probability_dict = {}
                        for k in node_dict:
                            nodes_edges = len(node_dict[k]) # This is the amount of edges a node has
                            probability_dict[k] = nodes_edges/total_edges # this makes a dictionary with the edge probability per node
                for j in range(1, self.m+1):
                    max_prob = max(probability_dict.values())
                    max_key = list(probability_dict.keys())[list(probability_dict.values()).index(max_prob)]
                    amount_of_maxes = sum(value == max_prob for value in probability_dict.values())
                    if amount_of_maxes == 1: 
                        node_dict[node_count].append(max_key)
                        node_dict[max_key].append(node_count)
                        del(probability_dict[max_key])
                    else:
                        new_prob_dict = {} # This is a new probability dictionary for when there are multiple highest values
                        for value in probability_dict.values():
                            if value == max_prob:
                                value_key = list(probability_dict.keys())[list(probability_dict.values()).index(max_prob)]
                                new_prob_dict[value_key] = max_prob
                                probability_dict[value_key] = 0
                            else:
                                continue
                        choose_node = random.choice(list(new_prob_dict.keys()))
                        node_dict[node_count].append(choose_node)
                        node_dict[choose_node].append(node_count)
                        del(probability_dict[choose_node])
                node_count += 1
        return node_dict
    
class Graph():
    """The function requires to specify a strategy and requires an amount of nodes.
    The class uses the nodes and strategy classes to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = {}
        self.strategy_dict = {}
    def generate(self):
        node_initialization = Nodes()
        generate_nodes = node_initialization.create_nodes(self.n_nodes)
        self.strategy_dict = self.strategy.edges(self.n_nodes, generate_nodes)
        return self.strategy_dict
    def __str__(self):
        return "Graph: {}".format(self.strategy_dict.items())

z = Graph(BarabasiAlbertUndirected(3),5)
z.generate()
print(z)

Graph: dict_items([(1, [2, 4, 5, 6]), (2, [1, 3, 4, 5, 6, 7, 8]), (3, [2, 4, 5, 6, 7, 8]), (4, [2, 1, 3]), (5, [2, 3, 1]), (6, [2, 3, 1, 7, 8]), (7, [2, 3, 6]), (8, [2, 3, 6])])


### D:
The respective graph generation strategies should check that their arguments make
sense (error checks).

In [69]:
import random
import timeit

class Nodes():
    """The class takes a number of nodes as argument.
    It then creates a dictionary containing the number of nodes as keys and empty lists as values.
    The empty lists will later be used to add the edges
    """
    def create_nodes(self, n_nodes):
        nodes = {}
        for i in range(1, n_nodes+1):
            nodes[i] = []
        return nodes

class ErdosRenyiDirected():
    """The class takes a probability as argument.
    It uses the dictionary from the nodes class and adds the edges (values), if the probability threshold is exceeded,
    to the nodes (keys). 
    The graph is directed, so and edge between node 1 and 2 does not mean and edge between node 2 and 1.
    """
    def __init__(self, probability):
        self.probability = probability
    def edges(self, n_nodes, node_dict):
        if self.probability > 1 or self.probability <0:
            raise ValueError("Probability cannot be more than 1.0")
        if n_nodes < 0:
            raise ValueError("Amount of nodes cannot be negative")
        for i in range(1,n_nodes+1):
            for j in range(1, n_nodes+1):
                edge_prob = random.uniform(0, 1)
                if edge_prob > self.probability:
                    continue
                else:
                    if j == i:
                        continue
                    else:
                        node_dict[i].append(j)
        return node_dict
    
class ErdosRenyiUndirected():
    """The class takes a probability as argument.
    It uses the dictionary from the nodes class and adds the edges (values), if the probability threshold is exceeded,
    to the nodes (keys). 
    The graph is undirected, so and edge between node 1 and 2 does automatically generates and edge between node 2 and 1.
    """
    def __init__(self, probability):
        self.probability = probability
    def edges(self, n_nodes, node_dict):
        if self.probability > 1 or self.probability <0:
            raise ValueError("Probability cannot be more than 1.0")
        if n_nodes < 0:
            raise ValueError("Amount of nodes cannot be negative")
        for i in range(1,n_nodes+1):
            for j in range(1, n_nodes+1):
                edge_prob = random.uniform(0, 1)
                if edge_prob > self.probability:
                    continue
                else:
                    if j == i:
                        continue
                    else:
                        if i in node_dict[i] or j in node_dict[i]:
                            continue
                        else:
                            node_dict[i].append(j)
                            node_dict[j].append(i)
        return node_dict

class BarabasiAlbertUndirected():
    """The function takes an amount of nodes to be added and an amount of edges (m) to be added for each new node.
    The graph is initialized with three nodes and some edges. For each new node and each new edge m: the node to which
    said edge is connected is determined by the amount of edges a node has with respect to the total amount of edges.
    If there is a 'dominant' node (most edges), then an edge is made between the new node and said dominant node. 
    If there are more then one nodes with the same amount of edges, chance decides with node gets an edge with the new node.
    """
    def __init__(self, m):
        self.m = m
    def edges(self, n_nodes, useless_dict):
        if self.m < 1:
            raise ValueError("M must be postive")
        if self.m > 3:
            raise ValueError("M cannot be more than 3, since there are only three nodes at initialization")
        if n_nodes < 1:
            raise ValueError("Amount of added nodes must be positive")
        node_count = 4 #Starting number, since there are 3 inital nodes
        node_dict = {1:[2], 2:[1,3], 3:[2]} # the initial setup
        while node_count <= n_nodes+3:
                node_dict[node_count] = []
                for i in node_dict:
                    if i == node_count:
                        continue
                    else:
                        prob = random.uniform(0, 1)
                        total_edges = 0
                        for j in node_dict:
                            total_edges += len(node_dict[j])
                        probability_dict = {}
                        for k in node_dict:
                            nodes_edges = len(node_dict[k]) # This is the amount of edges a node has
                            probability_dict[k] = nodes_edges/total_edges # this makes a dictionary with the edge probability per node
                for j in range(1, self.m+1):
                    max_prob = max(probability_dict.values())
                    max_key = list(probability_dict.keys())[list(probability_dict.values()).index(max_prob)]
                    amount_of_maxes = sum(value == max_prob for value in probability_dict.values())
                    if amount_of_maxes == 1: 
                        node_dict[node_count].append(max_key)
                        node_dict[max_key].append(node_count)
                        del(probability_dict[max_key])
                    else:
                        new_prob_dict = {} # This is a new probability dictionary for when there are multiple highest values
                        for value in probability_dict.values():
                            if value == max_prob:
                                value_key = list(probability_dict.keys())[list(probability_dict.values()).index(max_prob)]
                                new_prob_dict[value_key] = max_prob
                                probability_dict[value_key] = 0
                            else:
                                continue
                        choose_node = random.choice(list(new_prob_dict.keys()))
                        node_dict[node_count].append(choose_node)
                        node_dict[choose_node].append(node_count)
                        del(probability_dict[choose_node])
                node_count += 1
        return node_dict
    
class Graph():
    """The function requires to specify a strategy and requires an amount of nodes.
    The class uses the nodes and strategy classes to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = {}
        self.strategy_dict = {}
    def generate(self):
        node_initialization = Nodes()
        generate_nodes = node_initialization.create_nodes(self.n_nodes)
        self.strategy_dict = self.strategy.edges(self.n_nodes, generate_nodes)
        return self.strategy_dict
    def __str__(self):
        return "Graph: {}".format(self.strategy_dict.items())

### E:
Optimize your graph generation strategies in at least two different methodological ways
(e.g. using functional programming and Numba), and test the impact (using the timeit
function). Document that in your IPython Notebook.


In [1]:
"""
import sys
!{sys.executable} -m pip install numba
#Still can't get numba working

"""

"\nimport sys\n!{sys.executable} -m pip install numba\n#Still can't get numba working\n\n"

In order to make the program more functional, I first had to change the way it worked by replacing dictionaries with lists.
The graph is still represented as a dictionary, which is formed in the __str__ method upon generation. I introduced a seperate function for the error check. In addition, I used some list comprehension in the code. However, I was not able to add more functionallity to the code.

In [70]:
import random 

class Nodes():
    """The class takes a number of nodes as argument.
    It then creates a list containing the number of nodes.
    """
    def __init__(self):
        self.nodes = []
    def error_check(self, n_nodes):
        if n_nodes < 1:
            raise ValueError("Amount of nodes must be positive")
    def create_nodes(self, n_nodes):
        [self.nodes.append(i) for i in range(0,n_nodes)]
        return self.nodes
    def __str__(self):
        return '{}'.format(self.nodes)


class ErdosRenyiDirected():
    """The class takes a probability as argument.
    It uses the dictionary from the nodes class and adds the edges (values), if the probability threshold is exceeded,
    to the nodes (keys). 
    The graph is directed, so and edge between node 1 and 2 does not mean and edge between node 2 and 1.
    """
    def __init__(self, probability):
        self.probability = probability
        self.edge_list = []
    def error_check(self):
        if self.probability > 1 or self.probability <0:
            raise ValueError("Probability cannot be more than 1.0")
    def edges(self, n_nodes, node_list):
        self.edge_list = [[] for x in range(n_nodes)]
        for i in range(0,n_nodes):
            for j in range(0, n_nodes):
                edge_prob = random.uniform(0, 1)
                if edge_prob > self.probability:
                    continue
                else:
                    if j == i:
                        continue
                    else:  
                        self.edge_list[i].append(j)
        return node_list, self.edge_list
    def __str__(self):
        return '{}'.format(self.edge_list)

class ErdosRenyiUndirected():
    """The class takes a probability as argument.
    It uses the dictionary from the nodes class and adds the edges (values), if the probability threshold is exceeded,
    to the nodes (keys). 
    The graph is undirected, so and edge between node 1 and 2 does automatically generates and edge between node 2 and 1.
    """
    def __init__(self, probability):
        self.probability = probability
        self.edge_list = []
    def error_check(self):
            if self.probability > 1 or self.probability <0:
                raise ValueError("Probability cannot be more than 1.0")
    def edges(self, n_nodes, node_list):
        self.edge_list = [[] for x in range(n_nodes)]
        for i in range(0,n_nodes):
            for j in range(0, n_nodes):
                edge_prob = random.uniform(0, 1)
                if edge_prob > self.probability:
                    continue
                else:
                    if j == i:
                        continue
                    else:  
                        if i in self.edge_list[j] or j in self.edge_list[i]:
                            continue
                        else:
                            self.edge_list[i].append(j)
                            self.edge_list[j].append(i)
        return node_list, self.edge_list
    def __str__(self):
        return '{}'.format(self.edge_list)

class BarabasiAlbertUndirected():
    """The function takes an amount of nodes to be added and an amount of edges (m) to be added for each new node.
    The graph is initialized with three nodes and some edges. For each new node and each new edge m: the node to which
    said edge is connected is determined by the amount of edges a node has with respect to the total amount of edges.
    If there is a 'dominant' node (most edges), then an edge is made between the new node and said dominant node. 
    If there are more then one nodes with the same amount of edges, chance decides with node gets an edge with the new node.
    """
    def __init__(self, m):
        self.m = m
        self.edge_list = []
    def error_check(self):
            if self.m < 1:
                raise ValueError("M must be postive")
            if self.m > 3:
                raise ValueError("M cannot be more than 3, since there are only three nodes at initialization")
    def edges(self, n_nodes, node_list):
        self.edge_list = [[] for x in range(n_nodes+3)] #Plus three because there are three initial nodes
        for i in range(len(node_list), len(node_list)+3):
            node_list.append(i)
        node_count = 3 #Starting number, since there are 3 inital nodes
        self.edge_list = [[] for x in range(n_nodes+3)]
        self.edge_list[0].append(1) #These are the edges presence at initialization
        self.edge_list[1].append(0)
        self.edge_list[1].append(2)
        self.edge_list[2].append(1)
        while node_count <= n_nodes+2:
                for i in node_list:
                    if i == node_count:
                        continue
                    else:
                        prob = random.uniform(0, 1)
                        total_edges = 0
                        for j in self.edge_list:
                            total_edges += len(j)
                        probability_dict = {}
                        prob_index = 0
                        for k in self.edge_list:
                            nodes_edges = len(k) # This is the amount of edges a node has
                            probability_dict[prob_index] = nodes_edges/total_edges # this makes a dictionary with the edge probability per node
                            prob_index +=1 
                for j in range(1, self.m+1):
                    max_prob = max(probability_dict.values())
                    max_key = list(probability_dict.keys())[list(probability_dict.values()).index(max_prob)]
                    amount_of_maxes = sum(value == max_prob for value in probability_dict.values())
                    if amount_of_maxes == 1: 
                        self.edge_list[node_count].append(max_key)
                        self.edge_list[max_key].append(node_count)
                        del(probability_dict[max_key])
                    else:
                        new_prob_dict = {} # This is a new probability dictionary for when there are multiple highest values
                        for value in probability_dict.values():
                            if value == max_prob:
                                value_key = list(probability_dict.keys())[list(probability_dict.values()).index(max_prob)]
                                new_prob_dict[value_key] = max_prob
                                probability_dict[value_key] = 0
                            else:
                                continue
                        choose_node = random.choice(list(new_prob_dict.keys())) 
                        self.edge_list[node_count].append(choose_node)
                        self.edge_list[choose_node].append(node_count)
                        del(probability_dict[choose_node])
                node_count += 1
        return node_list, self.edge_list

class Graph():
    """The function requires to specify a strategy and requires an amount of nodes.
    The class uses the nodes and edges strategy to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = []
        self.strategy_list = []
        self.return_dict = {}
        self.node_initialization = Nodes()
    def error_check(self):
        check1 = self.node_initialization.error_check(self.n_nodes)
        check2 = self.strategy.error_check()
    def generate(self):
        self.generate_nodes = self.node_initialization.create_nodes(self.n_nodes)
        self.strategy_list = self.strategy.edges(self.n_nodes, self.generate_nodes)
        return self.strategy_list
    def __str__(self): 
        node_list = self.strategy_list[0]
        edge_list = self.strategy_list[1]
        for i in node_list:
            self.return_dict[i] = edge_list[i] 
        return "Graph: {}".format(self.return_dict.items())

### F:
Provide unit tests for the classes (instantiation), and the graph generators, where you
test for typical cases and difficult border cases.

In [71]:
import unittest

class NodesTest(unittest.TestCase, Nodes):
    def test_error_check(self):
        """This checks whether giving an invalid value for n_nodes raises a ValueError."""
        n = Nodes()
        with self.assertRaises(ValueError):
            n.error_check(-1)
    def test_create_nodes(self):
        """These test check if the create nodes function returns the desired output."""
        n = Nodes()
        self.assertEqual(n.create_nodes(4), [0,1,2,3])
        n = Nodes()
        self.assertEqual(n.create_nodes(0), [])
        
class ErdosRenyiDirectedTest(unittest.TestCase, ErdosRenyiDirected):
    def test_error_check(self):
        """This checks whether giving an invalid value for probability raises a ValueError."""
        x = ErdosRenyiDirected(1.1)
        with self.assertRaises(ValueError):
            x.error_check()
        x = ErdosRenyiDirected(-0.1)
        with self.assertRaises(ValueError):
            x.error_check()
            
class ErdosRenyiUndirectedTest(unittest.TestCase, ErdosRenyiUndirected):
    def test_error_check(self):
        """This checks whether giving an invalid value for probability raises a ValueError."""
        y = ErdosRenyiUndirected(1.1)
        with self.assertRaises(ValueError):
            y.error_check()
        y = ErdosRenyiUndirected(-0.1)
        with self.assertRaises(ValueError):
            y.error_check()
            
class BarabasiAlbertUndirectedTest(unittest.TestCase, BarabasiAlbertUndirected):
    def test_error_check(self):
        """This checks whether giving invalid values for m raises an exception.
        m should be a postive number (>0) and cannot be higher than 3, since at initialization there are 3 nodes.
        Therefore, the first new node could never have an edge with 4 other nodes."""
        z = BarabasiAlbertUndirected(4)
        with self.assertRaises(ValueError):
            z.error_check()
        z = BarabasiAlbertUndirected(0)
        with self.assertRaises(ValueError):
            z.error_check()
    def test_edges(self):
        """ The first test checks the initialization. The second parameter for edges is an empty list, 
        since passing n_nodes = 0 through Nodes() results in an empty list.
        The second test checks if the result of adding one node with two edges results:
        1. In a node list with 4 nodes.
        2. In an edge list with the last node having two edges.
        """
        z = BarabasiAlbertUndirected(1)
        self.assertEqual(z.edges(0, []), ([0,1,2], [[1], [0,2], [1]]))
        z = BarabasiAlbertUndirected(2)
        result = z.edges(1, [0])
        self.assertEqual(result[0], [0,1,2,3])
        self.assertEqual(len(result[1][3]), 2)
    
if __name__ == '__main__':
    unittest.main(argv=['first-arg-is-ignored'], exit=False)

......
----------------------------------------------------------------------
Ran 6 tests in 0.051s

OK


There are no unit tests for the edge functions from ErdosRenyiDirected and ErdosRenyiUndirected.
This is because the outcome is unpredictable. There is no telling what the result will be, nor the length of the resulting lists.

## 3. Smart Graph Data Representations

### A:

Implement a graph data representation (class) using a
“normal” (non-sparse) array/list/matrix

In [72]:
import numpy as np

class Graph():
    """The function requires to specify a strategy, an amount of nodes and a representation type.
    The class uses the nodes and strategy classes to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes, rep_type):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = []
        self.strategy_list = []
        self.return_dict = {}
        self.node_initialization = Nodes()
        self.rep_type = rep_type
    def error_check(self):
        check1 = self.node_initialization.error_check(self.n_nodes)
        check2 = self.strategy.error_check()
    def generate(self):
        self.generate_nodes = self.node_initialization.create_nodes(self.n_nodes)
        self.strategy_list = self.strategy.edges(self.n_nodes, self.generate_nodes)
        return self.strategy_list
    def __str__(self): 
        representation_return = self.rep_type.represent(self.strategy_list[0], self.strategy_list[1])
        representation_return = str(representation_return)
        return representation_return
    
class ArrayRepresentation():
    """The class takes the node list and edge list from the graph function and represents it as a numpy array, with the correct shape"""
    def represent(self, node_list, edge_list):
        node_len = len(node_list)
        matrix_size = node_len*node_len
        array_list = []
        for i in range(0,matrix_size):
            array_list.append(0)
        edge_array = np.array(array_list)
        edge_array = edge_array.reshape(node_len, node_len)
        for i in node_list:
            for j in edge_list[i]:
                edge_array[i,j] = 1
        return edge_array
            
x = Graph(ErdosRenyiDirected(0.5), 7, ArrayRepresentation())
x.error_check()
x.generate()
print(x)
print()

y = Graph(ErdosRenyiUndirected(0.3), 10, ArrayRepresentation())
y.error_check()
y.generate()
print(y)
print()

z = Graph(BarabasiAlbertUndirected(3),5, ArrayRepresentation())
z.error_check()
z.generate()
print(z)

[[0 0 1 0 1 1 0]
 [1 0 1 0 1 0 1]
 [1 1 0 1 0 1 1]
 [1 0 0 0 1 1 0]
 [1 1 1 0 0 1 1]
 [0 0 1 0 0 0 1]
 [1 1 1 0 0 1 0]]

[[0 0 1 1 1 1 0 0 0 1]
 [0 0 0 1 0 1 1 1 0 1]
 [1 0 0 1 0 0 1 0 0 0]
 [1 1 1 0 1 1 1 1 0 0]
 [1 0 0 1 0 0 1 1 1 0]
 [1 1 0 1 0 0 1 1 0 0]
 [0 1 1 1 1 1 0 0 1 1]
 [0 1 0 1 1 1 0 0 1 0]
 [0 0 0 0 1 0 1 1 0 0]
 [1 1 0 0 0 0 1 0 0 0]]

[[0 1 0 1 1 1 1 1]
 [1 0 1 1 0 0 0 0]
 [0 1 0 1 1 0 1 1]
 [1 1 1 0 1 1 1 1]
 [1 0 1 1 0 1 0 0]
 [1 0 0 1 1 0 0 0]
 [1 0 1 1 0 0 0 0]
 [1 0 1 1 0 0 0 0]]


Implement a graph data representation (class) using the sparse matrix from Numpy.


In [73]:
from scipy import sparse

class Graph():
    """The function requires to specify a strategy, an amount of nodes and a representation type.
    The class uses the nodes and strategy classes to generate the graph and then returns it.
    """
    def __init__(self, strategy, n_nodes, rep_type):
        self.strategy = strategy
        self.n_nodes = n_nodes
        self.generate_nodes = []
        self.strategy_list = []
        self.return_dict = {}
        self.node_initialization = Nodes()
        self.rep_type = rep_type
    def error_check(self):
        check1 = self.node_initialization.error_check(self.n_nodes)
        check2 = self.strategy.error_check()
    def generate(self):
        self.generate_nodes = self.node_initialization.create_nodes(self.n_nodes)
        self.strategy_list = self.strategy.edges(self.n_nodes, self.generate_nodes)
        return self.strategy_list
    def __str__(self): 
        representation_return = self.rep_type.represent(self.strategy_list[0], self.strategy_list[1])
        representation_return = str(representation_return)
        return representation_return
            
class SparseMatrixRepresentation():
    """The class takes the node list from the graph function as arguments.
    It then finds al the coordinates on which a value (edge) is stored and devides these over three arrays:
    a row, colum and the edge value (always 1). It then creates a scipy sparse matrix with coordinate format."""
    def represent(self, node_list, edge_list):
        node_len = len(node_list)
        matrix_size = node_len*node_len
        row_list = []
        col_list = []
        edges = []
        for i in node_list:
            for j in edge_list[i]:
                row_list.append(i)
                col_list.append(j)
                edges.append(1)
        rows = np.array(row_list)
        cols = np.array(col_list)
        edges2 = np.array(edges)
        sparse_matrix = sparse.coo_matrix((edges2, (rows, cols)))
        return sparse_matrix
    
x = Graph(ErdosRenyiDirected(0.5), 7, SparseMatrixRepresentation())
x.error_check()
x.generate()
print(x)
print()

y = Graph(ErdosRenyiUndirected(0.3), 10, SparseMatrixRepresentation())
y.error_check()
y.generate()
print(y)
print()

z = Graph(BarabasiAlbertUndirected(3),5, SparseMatrixRepresentation())
z.error_check()
z.generate()
print(z)

  (0, 3)	1
  (1, 0)	1
  (1, 3)	1
  (1, 4)	1
  (1, 5)	1
  (1, 6)	1
  (2, 1)	1
  (3, 0)	1
  (3, 2)	1
  (3, 4)	1
  (3, 6)	1
  (4, 1)	1
  (4, 3)	1
  (4, 5)	1
  (4, 6)	1
  (5, 2)	1
  (5, 4)	1
  (5, 6)	1
  (6, 0)	1
  (6, 2)	1
  (6, 4)	1

  (0, 2)	1
  (1, 2)	1
  (1, 4)	1
  (1, 9)	1
  (2, 1)	1
  (2, 0)	1
  (2, 5)	1
  (2, 6)	1
  (2, 9)	1
  (3, 5)	1
  (3, 8)	1
  (3, 9)	1
  (3, 6)	1
  (4, 1)	1
  (4, 7)	1
  (4, 5)	1
  (5, 2)	1
  (5, 3)	1
  (5, 4)	1
  (5, 6)	1
  (5, 7)	1
  (5, 8)	1
  (5, 9)	1
  (6, 2)	1
  (6, 5)	1
  (6, 3)	1
  (6, 8)	1
  (7, 4)	1
  (7, 5)	1
  (7, 9)	1
  (8, 3)	1
  (8, 5)	1
  (8, 6)	1
  (9, 3)	1
  (9, 1)	1
  (9, 2)	1
  (9, 5)	1
  (9, 7)	1

  (0, 1)	1
  (0, 3)	1
  (0, 4)	1
  (0, 5)	1
  (0, 6)	1
  (0, 7)	1
  (1, 0)	1
  (1, 2)	1
  (1, 3)	1
  (1, 5)	1
  (1, 7)	1
  (2, 1)	1
  (2, 3)	1
  (2, 4)	1
  (3, 1)	1
  (3, 2)	1
  (3, 0)	1
  (3, 4)	1
  (3, 5)	1
  (3, 6)	1
  (3, 7)	1
  (4, 3)	1
  (4, 2)	1
  (4, 0)	1
  (4, 6)	1
  (5, 3)	1
  (5, 1)	1
  (5, 0)	1
  (6, 3)	1
  (6, 0)	1
  (6, 4)	1
  (7, 3)

### B:

Test these representations with different parameters/border cases (and document your
findings in your notebook), and check (and document) the results with respect to the
memory/space needed, and any runtime differences that you can find.

In [74]:
x1 = Graph(ErdosRenyiDirected(1), 10, ArrayRepresentation())
x1.error_check()
x1.generate()
print(x1)
print()

x2 = Graph(ErdosRenyiDirected(1), 10, SparseMatrixRepresentation())
x2.error_check()
x2.generate()
print(x2)
print()

x3 = Graph(ErdosRenyiDirected(0.1), 10, ArrayRepresentation())
x3.error_check()
x3.generate()
print(x3)
print()

x4 = Graph(ErdosRenyiDirected(0.1), 10, SparseMatrixRepresentation())
x4.error_check()
x4.generate()
print(x4)
print()

x4 = Graph(ErdosRenyiDirected(0.5), 100, ArrayRepresentation())
x4.error_check()
x4.generate()
print(x4)
print()

x5 = Graph(ErdosRenyiUndirected(0.5), 100, SparseMatrixRepresentation())
x5.error_check()
x5.generate()
print(x5)
print()

x6 = Graph(BarabasiAlbertUndirected(1), 1, ArrayRepresentation())
x6.error_check()
x6.generate()
print(x6)
print()

x7 = Graph(BarabasiAlbertUndirected(3), 100, SparseMatrixRepresentation())
x7.error_check()
x7.generate()
print(x7)
print()

[[0 1 1 1 1 1 1 1 1 1]
 [1 0 1 1 1 1 1 1 1 1]
 [1 1 0 1 1 1 1 1 1 1]
 [1 1 1 0 1 1 1 1 1 1]
 [1 1 1 1 0 1 1 1 1 1]
 [1 1 1 1 1 0 1 1 1 1]
 [1 1 1 1 1 1 0 1 1 1]
 [1 1 1 1 1 1 1 0 1 1]
 [1 1 1 1 1 1 1 1 0 1]
 [1 1 1 1 1 1 1 1 1 0]]

  (0, 1)	1
  (0, 2)	1
  (0, 3)	1
  (0, 4)	1
  (0, 5)	1
  (0, 6)	1
  (0, 7)	1
  (0, 8)	1
  (0, 9)	1
  (1, 0)	1
  (1, 2)	1
  (1, 3)	1
  (1, 4)	1
  (1, 5)	1
  (1, 6)	1
  (1, 7)	1
  (1, 8)	1
  (1, 9)	1
  (2, 0)	1
  (2, 1)	1
  (2, 3)	1
  (2, 4)	1
  (2, 5)	1
  (2, 6)	1
  (2, 7)	1
  :	:
  (7, 2)	1
  (7, 3)	1
  (7, 4)	1
  (7, 5)	1
  (7, 6)	1
  (7, 8)	1
  (7, 9)	1
  (8, 0)	1
  (8, 1)	1
  (8, 2)	1
  (8, 3)	1
  (8, 4)	1
  (8, 5)	1
  (8, 6)	1
  (8, 7)	1
  (8, 9)	1
  (9, 0)	1
  (9, 1)	1
  (9, 2)	1
  (9, 3)	1
  (9, 4)	1
  (9, 5)	1
  (9, 6)	1
  (9, 7)	1
  (9, 8)	1

[[0 0 0 0 0 0 1 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 1 0 1 0 0]
 [0 0 0 1 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 1 0 1 0 1 0 0 0]
 [0 0 1 0 0 1 0 0 0 0]
 [0 

The program seems to deal fine with all the cases.

In [52]:
#checking the runtime
import sys

%time x1 = Graph(ErdosRenyiDirected(0.5), 7, ArrayRepresentation())
x1.error_check()
%time x1.generate()
print(x1)
print()

%time x2 = Graph(ErdosRenyiDirected(0.5), 7, SparseMatrixRepresentation())
x2.error_check()
%time x2.generate()
print(x2)
print()

%time y1 = Graph(ErdosRenyiUndirected(0.3), 10, ArrayRepresentation())
y1.error_check()
%time y1.generate()
print(y1)
print()

%time y2 = Graph(ErdosRenyiUndirected(0.3), 10, SparseMatrixRepresentation())
y2.error_check()
%time y2.generate()
print(y2)
print()

%time z1 = Graph(BarabasiAlbertUndirected(3),5, ArrayRepresentation())
z1.error_check()
%time z1.generate()
print(z1)

%time z2 = Graph(BarabasiAlbertUndirected(3),5, SparseMatrixRepresentation())
z2.error_check()
%time z2.generate()
print(z2)

CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 19.8 µs
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 57.9 µs
[[0 0 0 0 1 0 0]
 [1 0 1 0 0 0 1]
 [0 0 0 0 1 0 0]
 [1 0 1 0 0 0 1]
 [0 0 1 1 0 0 0]
 [0 1 0 1 0 0 1]
 [1 1 1 1 1 1 0]]

CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 15 µs
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 52.7 µs
  (0, 2)	1
  (0, 4)	1
  (0, 5)	1
  (0, 6)	1
  (1, 3)	1
  (1, 5)	1
  (1, 6)	1
  (2, 1)	1
  (2, 3)	1
  (2, 5)	1
  (3, 0)	1
  (3, 2)	1
  (3, 5)	1
  (5, 3)	1
  (6, 1)	1
  (6, 4)	1

CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 12.4 µs
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 95.6 µs
[[0 0 0 1 1 1 0 1 0 0]
 [0 0 0 1 0 0 1 1 0 0]
 [0 0 0 1 0 1 0 0 0 0]
 [1 1 1 0 1 0 0 1 1 1]
 [1 0 0 1 0 1 1 1 1 1]
 [1 0 1 0 1 0 1 1 0 0]
 [0 1 0 0 1 1 0 0 1 0]
 [1 1 0 1 1 1 0 0 0 1]
 [0 0 0 1 1 0 1 0 0 0]
 [0 0 0 1 1 0 0 1 0 0]]

CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 11.2 µs
CPU times: user 0 ns, sy

The runtime for the SpareMatrixRepresentation is consequently lower than that of the ArrayRepresentation. However, the difference is very small (about 8 µs). 

In [75]:
# Checking the memory needed 

class ArrayRepresentation():
    """The class takes the nodelist and edge list from the graph function and represents it as a numpy array, with the correct shape"""
    def represent(self, node_list, edge_list):
        node_len = len(node_list)
        matrix_size = node_len*node_len
        array_list = []
        for i in range(0,matrix_size):
            array_list.append(0)
        edge_array = np.array(array_list)
        edge_array = edge_array.reshape(node_len, node_len)
        for i in node_list:
            for j in edge_list[i]:
                edge_array[i,j] = 1
        x = sys.getsizeof(edge_array)
        return x

            
class SparseMatrixRepresentation():
    """The class takes the node list from the graph function as arguments.
    It then finds al the coordinates on which a value (edge) is stored and devides these over three arrays:
    a row, colum and the edge value (always 1). It then creates a scipy sparse matrix with coordinate format."""
    def represent(self, node_list, edge_list):
        node_len = len(node_list)
        matrix_size = node_len*node_len
        row_list = []
        col_list = []
        edges = []
        for i in node_list:
            for j in edge_list[i]:
                row_list.append(i)
                col_list.append(j)
                edges.append(1)
        rows = np.array(row_list)
        cols = np.array(col_list)
        edges2 = np.array(edges)
        sparse_matrix = sparse.coo_matrix((edges2, (rows, cols)))
        x = sys.getsizeof(sparse_matrix)
        return x
    
x1 = Graph(ErdosRenyiDirected(0.5), 7, ArrayRepresentation())
x1.error_check()
x1.generate()
print(x1)
print()

x2 = Graph(ErdosRenyiDirected(0.5), 7, SparseMatrixRepresentation())
x2.error_check()
x2.generate()
print(x2)
print()

y1 = Graph(ErdosRenyiUndirected(0.3), 10, ArrayRepresentation())
y1.error_check()
y1.generate()
print(y1)
print()

y2 = Graph(ErdosRenyiUndirected(0.3), 10, SparseMatrixRepresentation())
y2.error_check()
y2.generate()
print(y2)
print()

z1 = Graph(BarabasiAlbertUndirected(3),5, ArrayRepresentation())
z1.error_check()
z1.generate()
print(z1)

z2 = Graph(BarabasiAlbertUndirected(3),5, SparseMatrixRepresentation())
z2.error_check()
z2.generate()
print(z2)


112

56

112

56

112
56


The Array representation (112 bytes) requires double the memory size of the Sparse matrix (56 bytes).